src/eth1spec/rlp.py

```diff
@@ -14,270 +14,346 @@
 
 from __future__ import annotations
 
-from typing import List, Sequence, Tuple, Union, cast
+from typing import List, Sequence, Union, cast
 
 from .base_types import U256, Uint
 from .eth_types import Bytes
+from .utils.common import does_idx_exist
 
-verbose = False
-debug = False
+RLP = Union[Bytes, Uint, U256, Sequence["RLP"]]  # type: ignore
 
 
-RLP = Union[Bytes, Uint, U256, Sequence["RLP"]]  # type: ignore
+#
+# RLP Encode
+#
 
 
-def encode(x: RLP) -> Bytes:
+def encode(raw_data: RLP) -> Bytes:
     """
-    Encodes `x` into a sequence of bytes using RLP.
+    Encodes `raw_data` into a sequence of bytes using RLP.
 
     Parameters
     ----------
-    x : `RLP`
-        A `Bytes`, `Uint`, or sequence of `RLP` encodable objects.
+    raw_data : `RLP`
+        A `Bytes`, `Uint`, `Uint256` or sequence of `RLP` encodable
+        objects.
 
     Returns
     -------
     encoded : `eth1spec.eth_types.Bytes`
-        The RLP encoded bytes representing `x`.
+        The RLP encoded bytes representing `raw_data`.
     """
-    if verbose:
-        print("RLP(", x, ")", "type: ", type(x))
-    if isinstance(x, (bytearray, bytes)):
-        return R_b(x)
-    elif isinstance(x, (Uint, U256)):
-        return encode(x.to_be_bytes())
-    elif isinstance(x, Sequence):
-        return R_l(cast(Sequence[RLP], x))
+    if isinstance(raw_data, (bytearray, bytes)):
+        return encode_bytes(raw_data)
+    elif isinstance(raw_data, (Uint, U256)):
+        return encode(raw_data.to_be_bytes())
+    elif isinstance(raw_data, Sequence) and not isinstance(raw_data, str):
+        return encode_sequence(cast(Sequence[RLP], raw_data))
     else:
-        raise TypeError()
+        raise TypeError(
+            "RLP Encoding of type {} is not supported".format(type(raw_data))
+        )
 
 
-# binary encoding/decoding
-def R_b(x: Bytes) -> Bytes:
+def encode_bytes(raw_bytes: Bytes) -> Bytes:
     """
-    Encodes `x`, a sequence of bytes, using RLP.
+    Encodes `raw_bytes`, a sequence of bytes, using RLP.
 
     Parameters
     ----------
-    x : `eth1spec.eth_types.Bytes`
+    raw_bytes : `eth1spec.eth_types.Bytes`
         Bytes to encode with RLP.
 
     Returns
     -------
     encoded : `eth1spec.eth_types.Bytes`
-        The RLP encoded bytes representing `x`.
+        The RLP encoded bytes representing `raw_bytes`.
     """
-    if verbose:
-        print("R_b(", x, ")")
-    len_x = Uint(len(x))
-    if len_x == 1 and x[0] < 128:
-        return x  # bytearray([x[0] + 0x80]) # noqa: SC100
-    elif len_x < 56:
-        return bytearray([128 + len_x]) + x
+    len_raw_data = Uint(len(raw_bytes))
+
+    if len_raw_data == 1 and raw_bytes[0] < 128:
```

Comment:
If we use `MAX_SINGLE_BYTE`, then we would have to store `128 - 1` and also modify the code to replace `<` with `<=`.

I feel it would be hard to correctly name these constants. Looks like even geth [code](https://github.com/ethereum/go-ethereum/blob/7194c847b6e7f545f2aad57d8eae0a046e08d7a4/rlp/raw.go#L126) doesn't use names but rather the hex values.

I'll replace the numbers with the hex numbers. Let me know if I can drop the constants naming. Thanks.

Reply:
Yeah, I think hex is just fine!